File: scala/src/test/scala/com/shekhargulati/ninetynine_problems/_01_lists/P27Test.scala
package com.shekhargulati.ninetynine_problems._01_lists

import org.scalatest.{FunSpec, Matchers}

class P27Test extends FunSpec with Matchers {

  describe("P27 Spec part a") {
    it("should group 9 person into 3 groups of size 2,3, and 4") {
      val group = P27.group3(List("aldo", "beat", "carla", "david", "evi", "flip", "gary", "hugo", "ida"))
      group should have length 1260
    }
  }

  describe("P27 Spec part a second approach") {
    it("should group 9 person into 3 groups of size 2,3, and 4") {
      val group = P27.group3_1(List("aldo", "beat", "carla", "david", "evi", "flip", "gary", "hugo", "ida"))
      group should have length 1260
    }
  }

  describe("P27 Spec part b") {
    it("should group 9 person into 3 groups of size 2,3, and 4") {
      val group: List[List[String]] = P27.group(List("aldo", "beat", "carla", "david", "evi", "flip", "gary", "hugo", "ida"), List(2, 3, 4))
      group should have length 1260
    }

    it("should group 9 person into 3 groups of size 2,2, and 5") {
      val group: List[List[String]] = P27.group(List("aldo", "beat", "carla", "david", "evi", "flip", "gary", "hugo", "ida"), List(2, 2, 5))
      group should have length 756
    }
  }

  describe("P27 Spec part b second approach") {

    it("should group 3 person into 2 groups of size 1 and 2") {
      val group: List[List[String]] = P27.group_1(List("a", "b", "c"), List(1, 2))
      group should have length 3
    }


    it("should group 9 person into 3 groups of size 2,3, and 4") {
      val group: List[List[String]] = P27.group_1(List("aldo", "beat", "carla", "david", "evi", "flip", "gary", "hugo", "ida"), List(2, 3, 4))
      group should have length 1260
    }

    it("should group 9 person into 3 groups of size 2,2, and 5") {
      val group: List[List[String]] = P27.group_1(List("aldo", "beat", "carla", "david", "evi", "flip", "gary", "hugo", "ida"), List(2, 2, 5))
      group should have length 756
    }
  }

}

File: java8/src/test/java/com/shekhargulati/ninetynine_problems/_01_lists/P27Test.java
package com.shekhargulati.ninetynine_problems._01_lists;

import org.junit.Test;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;
import static org.hamcrest.collection.IsCollectionWithSize.hasSize;
import static org.junit.Assert.assertThat;

public class P27Test {

    @Test
    public void shouldGroupIntoThreeGroupsOfSize_2_3_and_4() throws Exception {
        List<String> input = Stream.of("aldo", "beat", "carla", "david", "evi", "flip", "gary", "hugo", "ida").collect(toList());
        List<List<List<String>>> groups = P27.group3(input);
        assertThat(groups, hasSize(1260));
    }

    @Test
    public void shouldGroupIntoThreeGroupsOfSize_2_2_and_5() throws Exception {
        List<String> input = Stream.of("aldo", "beat", "carla", "david", "evi", "flip", "gary", "hugo", "ida").collect(toList());
        List<List<List<String>>> groups = P27.group(input, Stream.of(2, 2, 5).collect(Collectors.toList()));
        assertThat(groups, hasSize(756));
    }

    @Test
    public void shouldGroupIntoThreeGroupsOfSize_1_and_2_part2() throws Exception {
        List<String> input = Stream.of("a", "b", "c").collect(toList());
        List<List<List<String>>> groups = P27.group(input, Stream.of(1, 2).collect(Collectors.toList()));
        assertThat(groups, hasSize(3));
    }


    @Test
    public void shouldGroupIntoThreeGroupsOfSize_2_3_and_4_part2() throws Exception {
        List<String> input = Stream.of("aldo", "beat", "carla", "david", "evi", "flip", "gary", "hugo", "ida").collect(toList());
        List<List<List<String>>> groups = P27.group(input, Stream.of(2, 3, 4).collect(Collectors.toList()));
        assertThat(groups, hasSize(1260));
    }

}